pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -183,6 +183,20 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
 	// Read the initial condition
 	before := pr.Status.GetCondition(apis.ConditionSucceeded)
 
+	// Check if we are failing to mark this as timed out for a while. If we are, mark immediately and finish the
+	// reconcile. We are assuming here that if the PipelineRun has timed out for a long time, it had time to run the
+	// before and it kept failing. One reason that can happen is exceeding etcd request size limit. Finishing it early
+	// makes sure the request size is manageable
+	if pr.HasTimedOutForALongTime(ctx, c.Clock) && !isPipelineRunTimeoutConditionSet(pr) {
+		if err := timeoutPipelineRun(ctx, logger, pr, c.PipelineClientSet); err != nil {
+			return err
+		}
+		if err := c.finishReconcileUpdateEmitEvents(ctx, pr, before, nil); err != nil {
```

Comment:
I'm a bit confused why this works. From the PR description and linked issue, it sounds like the controller is trying to update the pipelinerun status based on the taskrun status, and the request (which contains the full pipelinerun status) is too large to be handled by etcd. However, this calls finishReconcileUpdateEmitEvents, which calls updateLabelsAndAnnotations, which updates the pr with its current status. Why would this request not get stuck for being too big?

Is the idea that we skip including info from the child TaskRun that would make the status too large? If this is the case, it seems like this fix would only help for embedded-status=full (or both), since the childReferences are added as soon as the taskruns are created, and don't increase in size when the taskruns complete.

Reply:
Yes, I was thinking about embedded-status=full when implementing this.. but I think this scenario you mention would be covered too. If we receive a `request is too large` error while adding `childReferences`, wouldn't the PipelineRun get stuck in the `running` state in the cluster too? I believe this would prevent that too.

Even if the `updateLabelsAndAnnotations` fails with `request is too large`, I believe that wouldn't be a problem. Sure, the reconciliation would fail, but the UpdateStatus request would succeed in making the PipelineRun condition transition, so the PipelineRun wouldn't be stuck in a running state. That's because the UpdateStatus only apply changes in the status field.